### PR TITLE
Implement API validation

### PR DESCRIPTION
**Pull request by tpdownes**
This commit completes the implementation of a new validator that tests whether Google Cloud APIs required by the blueprint are actually enabled within the projects used by the blueprint.

An edge case to consider:

- if the project exists and the Compute Engine API is disabled, then validators for `[ test_project_exists, test_region_exists, test_zone_exists, test_zone_in_region]` will all fail because they use the Compute Engine API in their implementation
- if the project does not exist or the user cannot access it then `test_apis_enabled` will fail because one cannot query for the enabled APIs within the project

Thus both `test_project_exists` and and `test_apis_enabled` have some claim to being run first since they can both detect essentially the same problems (disabled APIs, non-existent/inaccessible projects). I have resolved this by electing to run `test_project_exists` first and reporting a special error when it fails due to the Compute Engine API itself being disabled. There is no way to determine that a project exists but you don't have access to it (hopefully obvious security reasons behind this).

Likewise, `test_apis_enabled` must report specially when either
- the Service Usage API is disabled (it queries this API to know which APIs are enabled)
- the project wasn't found or doesn't exist

If `test_project_exists` fails, all further validation stops (a current behavior that I have not changed). If the project wasn't found, then there isn't a lot to be done except print a message indicating that the project either doesn't exist or the user doesn't have permission. Thus, a user in the "maximally bad" scenario when validating a blueprint an in **existing** project will encounter:

1. A warning indicating that they need to enable the Compute Engine API (`test_project_exists` fails)
2. A warning indicating that they need to enable the Service Usage API (`test_project_exists` succeeds, but now `test_apis_enabled` fails in its attempt to query the enabled APIs)
3. Potential warnings indicating further APIs to enable based upon the blueprint's requirements
4. Potential warnings about region / zone global variables

### Submission Checklist

* [X] Have you installed and run this change against pre-commit? (`pre-commit
  install`)
* [X] Are all tests passing? (`make tests`)
* [X] Have you written unit tests to cover this change?
* [X] Is unit test coverage still above 80%?
* [X] Have you updated all applicable documentation?
* [X] Have you followed the guidelines in our Contributing document?

### CODE REVIEW

**Comment by tpdownes** [comment]
I will not merge this until an upcoming PR updating the README is also approved.

**Comment by tpdownes** [comment]
Sorry for the force pushes, I noticed that a new function was silently modifying one of its inputs (sorting it). It would have no impact on functionality in its current use, but could have been a debugging headache if it were put to other purposes down the road.

File: pkg/config/validator_test.go [hunk 1]
```diff
@@ -268,6 +266,53 @@ func (s *MySuite) TestGetStringValue(c *C) {
...
+	}
+	c.Assert(maps.EqualFunc(map3, expectedMap, slices.Equal[string]), Equals, true)
+
+	// unexpected value (duplicates removed and reverse sorted)
+	unexpectedMap := map[string][]string{
+		"key1": reversedValues1,
+	}
+	c.Assert(maps.EqualFunc(map3, unexpectedMap, slices.Equal[string]), Equals, false)
+
+	// TEST: merge with additional key in 2nd map
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/validator_test.go:298]: Fixing.

File: pkg/config/validator_test.go [hunk 2]
```diff
@@ -268,6 +266,53 @@ func (s *MySuite) TestGetStringValue(c *C) {
...
+	map3 = mergeMaps(map1, map2)
+
+	// test the expected value (duplicates removed and sorted)
+	expectedMap = map[string][]string{
+		"key1": slices.Clone(expectedValues1),
+		"key2": slices.Clone(expectedValues2),
+	}
+	c.Assert(maps.EqualFunc(map3, expectedMap, slices.Equal[string]), Equals, true)
+
+	// TEST: merge with additional key in 1st map (expected value unchanged!)
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/validator_test.go:309]: Fixing.

File: pkg/config/validate.go [hunk 3]
```diff
@@ -305,9 +309,36 @@ func testInputList(function string, inputs map[string]interface{}, requiredInput
...
+	maps.Copy(dest, base)
+
+	// sort each value in dest to ensure all output is sorted in case that the
+	// new map does not contain all keys in base
+	for _, v := range dest {
+		slices.Sort(v)
+	}
+
+	destKeys := maps.Keys(dest)
+	for newProject, newRequiredApis := range new {
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/validate.go:326]: Addressed and function simplified by using Go-native behavior of accessing keys of a map that do not yet have a value.

File: pkg/config/expand.go [hunk 4]
```diff
@@ -624,6 +626,13 @@ func (dc *DeploymentConfig) addDefaultValidators() error {
 		dc.Config.Validators = append(dc.Config.Validators, v)
 	}
 
+	// it is safe to run this validator even if vars.project_id is undefined;
+	// it will likely fail but will do so helpfully to the user
+	dc.Config.Validators = append(dc.Config.Validators, validatorConfig{
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:631]: I'm not sure what you're asking here. If what you're getting at is that we might have multiple deployment variables (or hard-coded values) that correspond to multiple project IDs (e.g. a host project in a Shared VPC environment).

I think the challenge is implicitly/automatically adding validators for multi-project environments where `$(vars.project_id)` is insufficient. Rather than ask the user to construct the validators manually, we are asking the user to manually decorate their modules with `required_apis`(a map from project IDs to a list of APIs) for these advanced use cases. This decoration supports the use of variables in the key, such that you can specify it as:

```yaml
required_apis:
  $(vars.project_id):
  - storage.googleapis.com
  $(vars.other_project_id):
  - compute.googleapis.com
```

Does this address your concern?

[the remark this answers is not in the document]
Comment (heyealex) [reply pkg/config/expand.go:631]: Sorry I wasn't clear, though I appreciate the context, it does answer my question.

I was curious if the intent was to infer from the `project_id` setting of a module rather than from the deployment variable of the same name at some point in the future. This option works well, though it might be worth trying in the future to avoid extra manual effort from the user.

**Review by heyealex — APPROVED** [review]

File: pkg/config/expand.go [hunk 5]
```diff
@@ -585,19 +597,31 @@ func (dc *DeploymentConfig) expandVariables() {
...
+				mod.Settings,
 				dc.ModuleToGroup)
 			if err != nil {
 				log.Fatalf("expandVariables: %v", err)
 			}
+
+			// ensure that variable references to projects in required APIs are expanded
+			for projectID, requiredAPIs := range mod.RequiredApis {
+				if isDeploymentVariable(projectID) {
+					s, err := updateVariableType(projectID, varContext{blueprint: dc.Config}, make(map[string]int))
```
Comment (heyealex) [inline pkg/config/expand.go:617]: Since this should _always_ be a string, passing it into `handleVariable` should have the same effect.

Comment (tpdownes) [reply pkg/config/expand.go:617]: Done on HEAD

**Comment by tpdownes** [comment]
I addressed one comment and also modified the implementation of automatically imputing required_apis to use `$(vars.project_id)` syntax because `expand` will now properly convert it to internal `((x.y))` format. I prefer that the automated solution look more like the user's manual approach.

**Review by heyealex — APPROVED** [review]
